### PR TITLE
IBX-5119: Added logging for invalid content's locations setup

### PR DESCRIPTION
**Pull request by barw4**
| Question                                  | Answer
| ---------------------------------------- | ------------------
| **JIRA issue**                          | [IBX-5119](https://issues.ibexa.co/browse/IBX-5119)
| **Type**                                   | improvement
| **Target Ibexa version** | `v3.3`
| **BC breaks**                          | no

Added logging when a main location for content doesn't exist but that content has locations assigned according to https://github.com/ezsystems/ezplatform-rest/pull/100#pullrequestreview-1310926719.

#### Checklist:
- [x] Provided PR description.
- [x] Tested the solution manually.
- [x] Provided automated test coverage.
- [x] Checked that target branch is set correctly (master for features, the oldest supported for bugs).
- [x] Ran PHP CS Fixer for new PHP code (use `$ composer fix-cs`).
- [x] Asked for a review (ping `@ezsystems/engineering-team`).

### CODE REVIEW

File: eZ/Publish/Core/Repository/Mapper/ContentDomainMapper.php [hunk 1]
```diff
@@ -119,6 +127,17 @@ public function buildContentDomainObject(
 
         $versionInfo = $this->buildVersionInfoDomainObject($spiContent->versionInfo, $prioritizedLanguages);
 
+        $contentInfo = $versionInfo->getContentInfo();
+        $mainLocation = $contentInfo->getMainLocation();
+        if ($mainLocation === null && $this->locationHandler->countAllLocations() > 0) {
```
Comment (alongosz) [inline eZ/Publish/Core/Repository/Mapper/ContentDomainMapper.php]: This is not correct. You're counting here all Locations. So if Content A has no Locations at all, but Content B has one main properly set Location, this expression would still be true for Content A. You need to count Locations for a given Content.

File: eZ/Publish/API/Repository/Values/Content/ContentInfo.php [hunk 2]
```diff
@@ -214,4 +214,9 @@ public function getOwner(): User
     {
         return $this->owner;
     }
+
+    public function getMainLocationId(): ?int
+    {
+        return $this->mainLocationId;
+    }
```
Comment (alongosz) [inline eZ/Publish/API/Repository/Values/Content/ContentInfo.php:221]: Nice. Now let's deprecate magic property by modifying line 30 to state
```php
* @property-read int|null $mainLocationId @deprecated Use {@see ContentInfo::getMainLocationId} instead
```

**Review by alongosz — CHANGES_REQUESTED** [review]

File: eZ/Publish/Core/Repository/Mapper/ContentDomainMapper.php (hunk 1 above)
Comment (barw4) [reply eZ/Publish/Core/Repository/Mapper/ContentDomainMapper.php]: Oops, I thought this method was receiving an argument and that I have provided one (IDE was not highlighting this), silly mistake

File: eZ/Publish/Core/Repository/Mapper/ContentDomainMapper.php [hunk 3]
```diff
@@ -119,6 +127,19 @@ public function buildContentDomainObject(
 
         $versionInfo = $this->buildVersionInfoDomainObject($spiContent->versionInfo, $prioritizedLanguages);
 
+        $contentInfo = $versionInfo->getContentInfo();
+        $mainLocation = $contentInfo->getMainLocation();
+        $contentLocations = $this->locationHandler->loadLocationsByContent($contentInfo->id);
```
Comment (alongosz) [inline eZ/Publish/Core/Repository/Mapper/ContentDomainMapper.php]: This operation is unfortunately too heavy for the mapper. We need here new method for Persistence handler, so you can call:
```suggestion
        $contentLocationsCount = $this->locationHandler->countContentItemLocations($contentInfo->id);
```
It needs to be defined on Location Persistence Handler interface, handler implementation, handler cache implementation and have dedicated Gateway method.

**Review by alongosz — CHANGES_REQUESTED** [review]

File: eZ/Publish/Core/settings/repository/inner.yml [hunk 4]
```diff
@@ -190,6 +190,8 @@ services:
             $contentLanguageHandler: '@ezpublish.spi.persistence.language_handler'
             $fieldTypeRegistry: '@eZ\Publish\Core\FieldType\FieldTypeRegistry'
             $thumbnailStrategy: '@eZ\Publish\Core\Repository\Strategy\ContentThumbnail\ThumbnailChainStrategy'
+        calls:
+            - [setLogger, ['@?logger']]
```
Comment (Steveb-p) [inline eZ/Publish/Core/settings/repository/inner.yml:194]: Add `monolog.logger` tag, and maybe add a channel for it.

```suggestion
        calls:
            - [setLogger, ['@?logger']]
        tags:
            - { name: 'monolog.logger', channel: 'ibexa.core' }
        
```

File: eZ/Publish/Core/Repository/Mapper/ContentDomainMapper.php (hunk 3 above)
Comment (barw4) [reply eZ/Publish/Core/Repository/Mapper/ContentDomainMapper.php]: @alongosz apparently there was already a gateway method `countLocationsByContentId` present - https://github.com/ezsystems/ezplatform-kernel/blob/1.3/eZ/Publish/Core/Persistence/Legacy/Content/Location/Gateway.php#L333. I've just implemented it into handler - please, see https://github.com/ezsystems/ezplatform-kernel/pull/364/commits/ae63e9aa0c644d0e2e88e02c9ead4bf9cc216cba

**Review by konradoboza — APPROVED** [review]

**Comment by konradoboza** [comment]
+1, apart from failing code style.

File: eZ/Publish/Core/Repository/Mapper/ContentDomainMapper.php [hunk 5]
```diff
@@ -119,6 +127,19 @@ public function buildContentDomainObject(
...
 
+        $contentInfo = $versionInfo->getContentInfo();
+        $mainLocation = $contentInfo->getMainLocation();
+
+        // For performance reasons 'countLocationsByContent' is moved to if
+        if ($mainLocation === null && $this->locationHandler->countLocationsByContent($contentInfo->id) > 0) {
+            $this->logger->error(
+                sprintf(
+                    'Main location for content of ID = %d doesn\'t exist yet this content has locations assigned.',
+                    $contentInfo->id
```
Comment (alongosz) [inline eZ/Publish/Core/Repository/Mapper/ContentDomainMapper.php]: Isn't this an instance which already has `getId` method? @barw4

Comment (barw4) [reply eZ/Publish/Core/Repository/Mapper/ContentDomainMapper.php]: Yes, it has after previous PR, resolved in https://github.com/ezsystems/ezplatform-kernel/pull/364/commits/528400a5ab0ec3242c00c64538fd48c669836323

File: eZ/Publish/Core/Persistence/Cache/LocationHandler.php [hunk 6]
```diff
@@ -502,4 +502,14 @@ private function getCacheTranslationKey(array $translations = null, bool $useAlw
 
         return implode('|', $translations) . '|' . (int)$useAlwaysAvailable;
     }
+
+    /**
+     * {@inheritdoc}
+     */
```
Comment (alongosz) [inline eZ/Publish/Core/Persistence/Cache/LocationHandler.php]: No need to inherit doc, it's inherited by default (I know we have a lot of obsolete PHPDocs like that).

File: eZ/Publish/Core/Repository/Mapper/ContentDomainMapper.php (hunk 5 above)
Comment (alongosz) [reply eZ/Publish/Core/Repository/Mapper/ContentDomainMapper.php]: Well, mostly I meant this line, but fix at 134 is good too, missed it.
```suggestion
                    $contentInfo->getId()
```

**Review by alongosz — CHANGES_REQUESTED** [review]

**Review by alongosz — APPROVED** [review]

**Review by micszo — APPROVED** [review]
QA Approved on Ibexa Experience 3.3.32-dev.

**Pull request merged**